Supported public GitHub for VideoMode Plugin Feature

mxin512 opened the pull request:
- Updated VideoModePlugin insturctions README.md for BA and BA:con
- Updated plugin examples for single and multi screens support

mxin512 committed:
Supported public GitHub for VideoMode Plugin Feature

- Updated VideoModePlugin insturctions README.md for BA and BA:con
- Updated plugin examples for single and multi screens support

bhastings1019 commented on README.md:
```diff
@@ -1,20 +1,21 @@
-
-1. Add videomode_plugin.brs to BrightAuthor (File > Presentation Properties > Autorun)
-1. Add a plugin message wherever you want to send it: Advanced tab > Add Command > Send > Send Plugin Message, pick the plugin you added in step 1, and add the text of the the plugin message e.g. "brightsign!videomode!1920x1080x60i!444!10bit" (format below)
+SetSingleScreenVideoMode.brs for non-series 5 single screen players.
```
> SetSingleScreenVideoMode.brs for non-series 5 single screen players.

Can SetSingleScreenVideoMode.brs be used for Series 5 single screen players, or Series 5 Multiscreen players while only using single HDMI output? 

Channeling questions customers may directly / indirectly ask or find confusion with when determining "which plugin to use".

Similar set of questions for...
> SetMultiScreensVideoMode.brs for series 5 multi screens players.

Can use on series 5 multi screen with only 1 hdmi output configured?

mxin512 replied:
> Can SetSingleScreenVideoMode.brs be used for Series 5 single screen players, or Series 5 Multiscreen players while only using single HDMI output?

No, SetSingleScreenVideoMode returns a string, and series 5 players call setVideoMode(as Array) which requires a parameter to be a screen Array. I have some related info added to the plugin. 

> Can use on series 5 multi screen with only 1 hdmi output configured?

Yes, it is possible if you just configure 1 HDMI, e.g. modifying screenArray[0]

bhastings1019 replied:
With the above understanding, could we update the below note to indicate single or multi screen _series 5_ hardware is unimportant here, what is important is that the hardware is series 5. 

> SetMultiScreensVideoMode.brs for series 5 multi screens players.

SetMultiScreensVideoMode.brs is used for series 5 players. SetScreenModes is used by series 5 hardware to support 1 to many connected screens.

mingjun7 replied:
The way Ming commented is the current implementation. A heads up that bacon team is planning to change to use SetMode for series 5 players but only support 1 screen. i.e. Raptor and Strata will always use multiscreen function, but the other series5 models only support single screen so goes back to the legacy API.

What we end up might be, recommend different script depend on the bacon version used for publish.

mingjun7 replied:
I want to set a flag here with the above comment that this can be a potential break point for series5 single screen players when bacon and the plugin are using different data structure.

mxin512 replied:
Sounds good, thanks for the note. Once there are new players with new videoMode instructions, we can update this GitHub page along the way.

bhastings1019 commented on README.md:
```diff
@@ -1,20 +1,21 @@
+SetMultiScreensVideoMode.brs for series 5 multi screens players.
 
-The format of command is a set of values separated by exclamation characters:
+# Usage
 
-    brightsign!videomode!<resolution/frame rate>!<color space>!<color depth>
```
Should we still be supporting this approach? There may be beneficial use cases customers to support a "plugin message" approach for how to "set a video mode".

mxin512 replied:
Added the approach of Plugin Message for videomode back.

bhastings1019 commented on README.md:
```diff
@@ -1,20 +1,21 @@
+# Supporting Resources
+Confluence: 
+1. MultiScreen VideoMode Configuration: https://brightsign.atlassian.net/wiki/spaces/DOC/pages/1208025142/VideoMode+Multiscreen+Configuration
+2. Supported Video Modes: https://brightsign.atlassian.net/wiki/spaces/DOC/pages/370676833/Supported+Video+Modes
```
Minor knit: for this, may improve from hyperlinks. 

example:
1. [MultiScreen VideoMode Configuration](https://brightsign.atlassian.net/wiki/spaces/DOC/pages/1208025142/VideoMode+Multiscreen+Configuration)
2. [Supported Video Modes](https://brightsign.atlassian.net/wiki/spaces/DOC/pages/370676833/Supported+Video+Modes)

bhastings1019 commented on videomode_plugin.brs:
Hesitant to say we should remove this. 

For the same of this PR, maybe not remove it unless it's truly incorrect / wrong to document this plugin. 

Otherwise, there may be benefit to keeping this for those who want to configure videomodes for single / multi output models using a plugin message dynamically not a static plugin file.

mxin512 replied:
Added original videomode_plugin.brs back.

bhastings1019 commented on README.md:
```diff
@@ -1,20 +1,32 @@
-# Usage
+SetSingleScreenVideoMode.brs for non-series 5 single screen players.
+SetMultiScreensVideoMode.brs for series 5 multi screens players.
 
```
Add line item mentioning what videoModePlugin.brs is similarly to how the single and multiscreen plugins are documented.

mxin512 replied:
Added one-line item mentioning what videoMode_plugin.brs is.

mingjun7 commented on README.md:
```diff
@@ -1,20 +1,32 @@
+A BrightSign plugin that changes the video resolution, frame rate, color space and color depth.
 
-# Usage
+SetSingleScreenVideoMode.brs for non-series 5 single screen players.
```
> for non-series 5 single screen players

Probably better to say series4 and below.

mxin512 replied:
Fixed.

mingjun7 commented on README.md:
```diff
@@ -1,20 +1,32 @@
+
+# Plugin Message Usage
+
+1. Add videomode_plugin.brs to BrightAuthor or BrightAuthor:Connected (File > Presentation Properties > Autorun)
```
It looks a bit confusing here. SetSingleScreenVideoMode.brs and SetMultiScreensVideoMode.brs, v.s. videomode_plugin.brs are alternatives to apply video mode. But the way it's writing sounds like you'll need implement both scripts. 
Can you make it more clear only one is needed? Also, videomode_plugin.brs only support one screen video mode.

mxin512 replied:
Added one-line item mentioning what videoMode_plugin.brs is.

mingjun7 commented on SetMultiScreensVideoMode.brs:
```diff
@@ -0,0 +1,40 @@
+	vm = CreateObject("roVideoMode")
+	sm = vm.GetScreenModes()
+
+	sm[0].video_mode="3840x2160x29.97p:rblank:fullres"
```
Could you add some comments to give more instructions for the following:

- Where are the areas that user should modify
- There is a field "name" that is important. ATM, it's default by order 0->hdmi1, 1->hdmi2, etc. But any script can change to order. It's safer to include name in the example to prevent mismatch and unexpected behavior
- Clarify that number of array items depend on the number of hdmi outputs. e.g. XC2055 has 3 - hdmi1, hdmi2, typeC
- if want to turn an output to off, set enabled = false is enough. No need to add the other fields.
- if do not want to touch the existing settings of a screen, can skip setting it.

mxin512 replied:
Details doc instructions page added. Hope that solves your ask.

mingjun7 replied:
It doesn't look like covers everything.

- Where are the areas that user should modify => suggest have a comment(e.g. ' TODO) at the beginning of where user should touch

- There is a field "name" that is important. ATM, it's default by order 0->hdmi1, 1->hdmi2, etc. But any script can change to order. It's safer to include name in the example to prevent mismatch and unexpected behavior => Can we include "name" in the example here?

- Clarify that number of array items depend on the number of hdmi outputs. e.g. XC2055 has 3 - hdmi1, hdmi2, typeC => Doesn't look like the docs is clear, it only has example for XC4055.

- if want to turn an output to off, set enabled = false is enough. No need to add the other fields. => This is mentioned in the doc

- if do not want to touch the existing settings of a screen, can skip setting it. => not mentioned in the doc

mxin512 replied:
> Can we include "name" in the example here?

This one confuses me a little since we don't have this field included in the docs page, and the plugin was written based on our current docs.

So it sounds to me that you are suggesting a few clarifications on MultiScreen Configuration Docs or roVideoMode, which those detailed info/asks can be a DOCS ticket. I do not feel like it is best practice to document all those info in the Plugin. Would that make sense to create a DOCS ticket? @mingjun7 @bhastings1019

mingjun7 replied:
@mxin512 name is on the docs page. It's just not in the example. See
"name: On the XC platform, these are HDMI-1 through HDMI-4 but can differ across platforms."

I leave you guys to decide if it helps to add the "name" field. Just want to mention that the order can be different.

Sounds good. We can have a DOCS ticket to capture the details. I filed https://brightsign.atlassian.net/browse/DOCS-1165.

mxin512 replied:
Aha, okay. Thanks for the docs ticket. I think it makes sense to align the videomode plugin with our docs example.

mingjun7 replied:
No problem. Stephanie has update the docs. Feel free to take a look.

mingjun7 commented on SetSingleScreenVideoMode.brs:
```diff
@@ -0,0 +1,21 @@
+
+	print "setvideomode_ForSingleScreen"
+
+	'vm=CreateObject("roVideoMode")
```
Can you cleanup the commented lines here? 
If want to attach example usages, please add explanation what they are and when to use them.

mxin512 replied:
One-line explanation with doc added.

mingjun7 commented on SetSingleScreenVideoMode.brs:
```diff
@@ -0,0 +1,21 @@
+' Single Screen VideoMode Plugin Example
+' Plugin function name - setvideomode
+' Plugin return type - As String. See BrightSign docs for a list of supported video modes.
```
Maybe link the docs urls here.

mxin512 replied:
Link attached.

mingjun7 commented on SetMultiScreensVideoMode.brs:
```diff
@@ -0,0 +1,40 @@
+' Plugin return type - As Array: For series5 players, it is plugin's responsibility to 
+' return the same format as GetScreenModes() returns. Then the presentation autorun will
+' pick up your ScreenConfigList and set video modes for your screens.
+' Note: BrightAuthor:connected always sets :fullres on all of the videomodes for the XC5
```
This will soon no longer be true. We are working on to enable using without :fullres.

mxin512 replied:
Note removed.

mxin512 committed:
Supported public GitHub for VideoMode Plugin Feature

- Updated VideoModePlugin insturctions README.md for BA and BA:con
- Updated plugin examples for single and multi screens support

mingjun7 commented on README.md:
```diff
@@ -1,20 +1,33 @@
+A BrightSign plugin that changes the video resolution, frame rate, color space and color depth.
 
-# Usage
+SetSingleScreenVideoMode.brs is used for series 4 and below.
```
md file does not automatically add new line break. You need to use "\<br/>\" to indicate for a new line.
You can see the preview with the "Display the rich diff" button on the top right.

mingjun7 commented on SetMultiScreensVideoMode.brs:
```diff
@@ -0,0 +1,40 @@
+Function setvideomode(videoModeInputs, bsp)
+
+	print "setvideomode_ForMultiScreens"
+  ' See details regarding multiScreen configuration - [MultiScreen VideoMode Configuration](https://brightsign.atlassian.net/wiki/spaces/DOC/pages/1208025142/VideoMode+Multiscreen+Configuration)
```
Minor - can you align the white space?
Also the [url](https://xxx) syntax only works for markdown file. Can directly use url in brightscript.

mingjun7 commented on SetSingleScreenVideoMode.brs:
```diff
@@ -0,0 +1,23 @@
+	'vm.SetCustomModeline("Modeline 1280x720x30p 35.25  1280 1320 1440 1600  720 723 728 736 -hsync +vsync")
+	'return "custom"
+	
+	'return "3840x600x60p:rblank"
```
Instead of having a bunch of examples as commented, does it make more sense to point to https://brightsign.atlassian.net/wiki/spaces/DOC/pages/370672969/roVideoMode#roVideoMode-SetMode(modeAsString)AsBooleansetmode for the options user can set?

mingjun7 commented on README.md:
```diff
@@ -1,20 +1,33 @@
-1. Add a plugin message wherever you want to send it: Advanced tab > Add Command > Send > Send Plugin Message, pick the plugin you added in step 1, and add the text of the the plugin message e.g. "brightsign!videomode!1920x1080x60i!444!10bit" (format below)
+# VideoMode Plugin Usage
+
+Add SetSingleScreenVideoMode.brs or SetMultiScreensVideoMode.brs to BrightAuthor or BrightAuthor:Connected (Presentation Settings > Support Content > Video Mode Plugin)
```
Feels to me the first step should be update SetSingleScreenVideoMode.brs or SetMultiScreensVideoMode.brs with the user's screens settings.

mxin512 committed:
Supported public GitHub for VideoMode Plugin Feature

- Updated VideoModePlugin insturctions README.md for BA and BA:con
- Updated plugin examples for single and multi screens support
